drivers/flash/flash_stm32f7x.c

```diff
@@ -0,0 +1,249 @@
+
+#include "flash_stm32.h"
+
+#define STM32F7X_SECTOR_MASK		((u32_t) 0xFFFFFF07)
+
+bool flash_stm32_valid_range(struct device *dev, off_t offset, u32_t len,
+			     bool write)
+{
+	ARG_UNUSED(write);
+
+	return flash_stm32_range_exists(dev, offset, len);
+}
+
+static int write_byte(struct device *dev, off_t offset, u8_t val)
+{
+	struct stm32f7x_flash *regs = FLASH_STM32_REGS(dev);
+	int rc;
+
+	/* if the control register is locked, do not fail silently */
+	if (regs->cr & FLASH_CR_LOCK) {
+		return -EIO;
+	}
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	if (rc < 0) {
+		return rc;
+	}
+
+	regs->cr &= CR_PSIZE_MASK;
+	regs->cr |= FLASH_PSIZE_BYTE;
+	regs->cr |= FLASH_CR_PG;
+
+	/* flush the register write */
+	__DSB();
+
+	*((u8_t *) offset + CONFIG_FLASH_BASE_ADDRESS) = val;
+
+	/* flush the register write */
+	__DSB();
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	regs->cr &= (~FLASH_CR_PG);
+
+	return rc;
+}
+
+static int erase_sector(struct device *dev, u32_t sector)
+{
+	struct stm32f7x_flash *regs = FLASH_STM32_REGS(dev);
+	int rc;
+
+	/* if the control register is locked, do not fail silently */
+	if (regs->cr & FLASH_CR_LOCK) {
+		return -EIO;
+	}
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	if (rc < 0) {
+		return rc;
+	}
+/* Dual bank mode, SNB MSB selects the bank2,
+ * sothers select sector, So we remap sector number.
+ */
+#ifdef FLASH_OPTCR_nDBANK
+#if CONFIG_FLASH_SIZE == 1024
+	if (sector > 11) {
+		sector += 4;
+	}
+#elif CONFIG_FLASH_SIZE == 1024
+	if (sector > 7) {
+		sector += 8;
+	}
+#endif /* CONFIG_FLASH_SIZE */
+#endif /* defined(FLASH_OPTCR_nDBANK) */
+
+	regs->cr &= (STM32F7X_SECTOR_MASK & CR_PSIZE_MASK);
+	regs->cr |= FLASH_PSIZE_BYTE;
+	regs->cr |= FLASH_CR_SER | (sector << FLASH_CR_SNB_Pos);
+	regs->cr |= FLASH_CR_STRT;
+
+	/* flush the register write */
+	__DSB();
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	regs->cr &= ~(FLASH_CR_SER | FLASH_CR_SNB);
+
+	return rc;
+}
+
+int flash_stm32_block_erase_loop(struct device *dev, unsigned int offset,
+				 unsigned int len)
+{
+	struct flash_pages_info info;
+	u32_t start_sector, end_sector;
+	u32_t i;
+	int rc = 0;
+
+	rc = flash_get_page_info_by_offs(dev, offset, &info);
+	if (rc) {
+		return rc;
+	}
+	start_sector = info.index;
+	rc = flash_get_page_info_by_offs(dev, offset + len - 1, &info);
+	if (rc) {
+		return rc;
+	}
+	end_sector = info.index;
+
+	for (i = start_sector; i <= end_sector; i++) {
+		rc = erase_sector(dev, i);
+		if (rc < 0) {
+			break;
+		}
+	}
+
+	return rc;
+}
+
+int flash_stm32_write_range(struct device *dev, unsigned int offset,
+			    const void *data, unsigned int len)
+{
+	int i, rc = 0;
+
+	for (i = 0; i < len; i++, offset++) {
+		rc = write_byte(dev, offset, ((const u8_t *) data)[i]);
+		if (rc < 0) {
+			return rc;
+		}
+	}
+
+	return rc;
+}
+
+/* Different SoC flash layouts are specified in across various
+ * reference manuals, but the flash layout for a given number of
+ * sectors. In case of 8 sectors we need to differentiate between
```

Comment:
'but the flash layout for a given number of sectors.' Do we miss a word in this part of the sentence?